File: gui.java
import javax.swing.*;
import java.awt.*;
import java.util.*;
import java.io.*;
import java.time.*;
import java.awt.event.*;

class gui {
    private static ArrayList<JLabel> Labels = new ArrayList<>();
    private static ArrayList<Object> Text = new ArrayList<>();
    private static JComboBox<String> SliceDropdown;
    private static JComboBox<String> IntervalDropdown;
    private static JTextArea ta = new JTextArea();
    private static JPanel panel = new JPanel();
    private static JMenuBar mb = new JMenuBar();
    private static JFrame frame = new JFrame("The Price of Free Speech");
    private static String toPrint = "";
    private static boolean Resizable = true;
    private static Interaction interaction;
    private static Interaction DIA;

    public static void main(String[] args) {

        //Creating the Frame
        frame.setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
        frame.setSize(1000, 400);
        frame.setResizable(Resizable);

        //Creating the MenuBar and adding components
        JMenu file = new JMenu("FILE");
        JMenu help = new JMenu("Help");
        mb.add(file);
        mb.add(help);
        JMenuItem open = new JMenuItem("Open");
        JMenuItem save_as = new JMenuItem("Save as");
        file.add(open);
        file.add(save_as);
        JButton select = new JButton("Begin");
        createLabels();
        createText();
        createDropdown();
        select.addActionListener(new ActionListener(){
            @Override
            public void actionPerformed(ActionEvent e){
                try {
                    String ticker = ((JTextField)Text.get(0)).getText();
                    int IntervalInt = IntervalDropdown.getSelectedIndex();
                    int SliceInt = SliceDropdown.getSelectedIndex();
                    String handles = ((JTextField)Text.get(2)).getText();
                    interaction = new Interaction(ticker,IntervalInt,SliceInt,handles,false);
                    DIA = new Interaction("DIA",IntervalInt,SliceInt,handles,true);
                    writeDate();
                    interaction.run();
                    DIA.run();
                    writeSMVI();
                    ta.append("\n"+toPrint);
                } catch (IOException ex) {
                    ex.printStackTrace();
                }
            }
        });
        for(int i = 0; i < Labels.size(); i++){
            panel.add(Labels.get(i));
            if(i == 1){
                panel.add(IntervalDropdown);
            }
            else if(i == 3){
                System.out.println("Creating Slice Dropdown");
                panel.add(SliceDropdown);
            }
            else panel.add((JTextField)(Text.get(i)));
        }
        // Dropdown Menu selection
        panel.add(select);
        ta = new JTextArea(toPrint);

        // Text Area at the Center
        //Adding Components to the frame.
        frame.getContentPane().add(BorderLayout.SOUTH, panel);
        frame.getContentPane().add(BorderLayout.NORTH, mb);
        frame.getContentPane().add(BorderLayout.CENTER, ta);
        frame.setVisible(true);
    }

    private static void writeDate(){
        String input = SliceDropdown.getSelectedItem().toString();
        try{
            FileWriter fileWriter = new FileWriter("Data\\Time.txt", false); 
            PrintWriter printWriter = new PrintWriter(fileWriter, false);
            printWriter.flush();
            fileWriter.write(input);
            fileWriter.close();
            printWriter.close();
        }catch(Exception e){System.err.println(e);}
    }

    private static void createLabels(){
        JLabel TickerLabel = new JLabel("Ticker");
        JLabel IntervalLabel = new JLabel("\tInterval (Minutes)");
        JLabel HandleLabel = new JLabel("\tTwitter Handle");
        JLabel SliceLabel = new JLabel("\tMonth");
        Labels = new ArrayList<>(Arrays.asList(TickerLabel,IntervalLabel,HandleLabel,SliceLabel));
    }

    private static void createText() {
        JTextField TickerText = new JTextField(6); // accepts up to 6 characters
        JTextField HandleText = new JTextField(15);
        Object[] temp = new Object[]{TickerText,IntervalDropdown,HandleText,SliceDropdown};
        Text = new ArrayList<>(Arrays.asList(temp));
    }

    /**
     * A method that calculates which months and years should be showning the dropdown menu as well as there order
     */
    private static void createDropdown() {
        ArrayList<String> Months = new ArrayList<>();
        String[] months = new String[]{"January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December" };
        LocalDate currentdate = LocalDate.now();
        int currentyear = currentdate.getYear();
        String currentmonth = currentdate.getMonth().toString();
        int MonthNumber = 0;
        int monthsLeft = 24;
        for (int i = 0; i < 12; i++) {
            if (months[i].equalsIgnoreCase(currentmonth)) MonthNumber = i;
        }
        //Add months from the current year
        for(int i = MonthNumber; i > -1; i--){
            Months.add(months[i]+" "+currentyear);
            monthsLeft--;
        }
        currentyear -= 1;
        int index = 11;
        //Add months from the previous years
        while(monthsLeft > 0){
            if(index == -1){
                index = 11;
                currentyear -= 1;
            }
            Months.add(months[index]+" "+ currentyear);
            monthsLeft--;
            index--;
        }
        SliceDropdown = new JComboBox(Months.toArray());
        String[] Intervals = new String[]{"1","5","15","30","60"};
        IntervalDropdown = new JComboBox(Intervals);
    }

    public static void WriteText(String Text){ toPrint += "\n"+Text;}

    
    private static void writeSMVI(){
        try{
            RunPython.Run(0);
            RunPython.Run(1);
            File file = new File("Data\\SMVI_Data.txt");
            FileWriter writer = new FileWriter(file);
            String[] varNames = new String[]{"va","na","vb","nb","T","t"};
            double va = interaction.getVariation(interaction.getRawData());
            double na= interaction.getNumberOfDataPoints();
            double vb = DIA.getVariation(DIA.getRawData());
            double nb = DIA.getNumberOfDataPoints();
            double t = 31; //TODO: Get the time from the range
            double T = Double.parseDouble(RunPython.getOutput(1));
            double[] var = new double[]{va,na,vb,nb,t,T};
            System.out.println("Writing");
            for(int i = 0; i < var.length; i++){
                writer.write(varNames[i]+":"+var[i]+"\n");
            }
            writer.close();
            RunPython.Run(2);
            WriteText("SMVI: " + RunPython.getOutput(2));
        }catch(Exception e){System.out.println("Error in \"writeSMVI\"\n"+e);}
    }

}

File: Gui.py
from AutoComplete import predictor
from calendar import monthrange
from datetime import date
import PySimpleGUI
import csv
import os



class Gui:

    def __init__(self):
        self.Intervals, self.Months, self.window, self.complete, self.allTickers, self.non_parsed_tickers = ([] for i in range(6))
        self.run()

    def format(self, foo):
        toReturn = foo.strftime('%B'), foo.strftime('%Y')
        toReturn = ' '.join(toReturn)
        return toReturn

    def getTickers(self):
        with open(os.getcwd() + '/Data/tickers.csv') as file:
            reader = csv.reader(file)
            for line in reader:
                if line[0] != 'Symbol' or line[0].split('|')[0] != 'Symbol':
                    if '^' in line[0] or '^' in line[0].split('|')[0]:
                        continue
                    if '|' not in line[0]:
                        self.non_parsed_tickers.append(line[0])
                    else:
                        self.non_parsed_tickers.append(line[0].split('|')[0])
        self.allTickers = list(dict.fromkeys(self.non_parsed_tickers))
        self.allTickers.sort()

    def getDates(self):
        months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October",
                  "November", "December"]
        days = []
        today_date = date.today()
        latestMonth = today_date.strftime('%B')
        latestMonth_Index = months.index(latestMonth) + 1
        tempMonthIndex = latestMonth_Index
        year = today_date.year
        lastYear = []
        for i in range(12):
            if tempMonthIndex + i == 13: break
            lastYear.append(date(year - 1, tempMonthIndex + i, 1))
            days.append(monthrange(year - 1, tempMonthIndex + i)[1])
        tempMonthIndex = latestMonth_Index - 1
        thisYear = []
        for f in range(latestMonth_Index):
            thisYear.append(date(year, tempMonthIndex - f + 1, 1))
            days.append(monthrange(year, tempMonthIndex - f + 1)[1])
        thisYear.reverse()
        return lastYear + thisYear

    def update(self, key, text):
        self.window[key].update(text)

    def updateTickers(self):
        event, values = self.data()
        closest = self.complete.getClosest(values['Ticker'])
        self.window.FindElement('Ticker').Update(values=closest, value=values['Ticker'], size=(10, 10))

    def data(self):
        return self.window.read(timeout=30)

    def run(self):
        self.complete = predictor()
        self.getTickers()
        self.Intervals = ['1 Minute', '5 Minutes', '15 Minutes', '30 Minutes', '60 Minutes', '1 Day', '1 Week',
                          '1 Month']
        for month in self.getDates():
            self.Months.append(self.format(month))
        self.layout = [
            [PySimpleGUI.Text('Stock Ticker', size=(15, 1)),    PySimpleGUI.InputCombo(values=self.allTickers, key='Ticker', default_value='Ticker', size=(10, 1))],
            [PySimpleGUI.Text('Twitter Handles (Separated by a comma)', size=(15, 3)),  PySimpleGUI.Input("Handle", key='Handle', size=(10, 1))],
            [PySimpleGUI.Text('Data Interval', size=(15, 1)),   PySimpleGUI.Combo(self.Intervals, default_value=self.Intervals[3], key='Interval')],
            [PySimpleGUI.Text('Start Month', size=(15, 1)),   PySimpleGUI.Combo(self.Months, default_value=self.Months[0], key='StartMonth')],
            [PySimpleGUI.Text('End Month', size=(15, 1)),   PySimpleGUI.Combo(self.Months, default_value=self.Months[len(self.Months) - 1], key='EndMonth')],
            [PySimpleGUI.Text('SD:', size=(5, 1)), PySimpleGUI.Text(size=(10, 1), key='SD')],
            [PySimpleGUI.Text('SMVI:', size=(5, 1)), PySimpleGUI.Text(size=(10, 1), key='SMVI')],
            [PySimpleGUI.Button('Run'),   PySimpleGUI.Text('Please enter in your information', size=(28, 1), key='Terminal')],
            [PySimpleGUI.Button("Exit")]
        ]
        PySimpleGUI.theme('Reddit')
        self.window = PySimpleGUI.Window("The Price of Free Speech", self.layout)
